File: android/auto-mobile-sdk/src/main/kotlin/dev/jasonpearson/automobile/sdk/logging/AutoMobileLog.kt
package dev.jasonpearson.automobile.sdk.logging

import android.util.Log
import dev.jasonpearson.automobile.protocol.SdkLogEvent
import dev.jasonpearson.automobile.sdk.events.SdkEventBuffer
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Drop-in replacement for [android.util.Log] that also checks registered filters
 * and posts matching entries to the SDK event buffer.
 *
 * Zero overhead when no filters are registered: a single `filters.isEmpty()` check.
 * First-match semantics — only records once per log call.
 */
object AutoMobileLog {

  @Volatile private var buffer: SdkEventBuffer? = null
  @Volatile private var applicationId: String? = null
  private val filters = CopyOnWriteArrayList<CompiledLogFilter>()

  fun initialize(applicationId: String?, buffer: SdkEventBuffer) {
    this.applicationId = applicationId
    this.buffer = buffer
  }

  fun addFilter(
    name: String,
    tagPattern: Regex? = null,
    messagePattern: Regex? = null,
    minLevel: Int = Log.VERBOSE,
  ) {
    filters.add(CompiledLogFilter(name, tagPattern, messagePattern, minLevel))
  }

  fun removeFilter(name: String) {
    filters.removeAll { it.name == name }
  }

  fun clearFilters() {
    filters.clear()
  }

  fun v(tag: String, msg: String): Int {
    val result = Log.v(tag, msg)
    checkFilters(tag, msg, Log.VERBOSE)
    return result
  }

  fun v(tag: String, msg: String, tr: Throwable): Int {
    val result = Log.v(tag, msg, tr)
    checkFilters(tag, msg, Log.VERBOSE)
    return result
  }

  fun d(tag: String, msg: String): Int {
    val result = Log.d(tag, msg)
    checkFilters(tag, msg, Log.DEBUG)
    return result
  }

  fun d(tag: String, msg: String, tr: Throwable): Int {
    val result = Log.d(tag, msg, tr)
    checkFilters(tag, msg, Log.DEBUG)
    return result
  }

  fun i(tag: String, msg: String): Int {
    val result = Log.i(tag, msg)
    checkFilters(tag, msg, Log.INFO)
    return result
  }

  fun i(tag: String, msg: String, tr: Throwable): Int {
    val result = Log.i(tag, msg, tr)
    checkFilters(tag, msg, Log.INFO)
    return result
  }

  fun w(tag: String, msg: String): Int {
    val result = Log.w(tag, msg)
    checkFilters(tag, msg, Log.WARN)
    return result
  }

  fun w(tag: String, msg: String, tr: Throwable): Int {
    val result = Log.w(tag, msg, tr)
    checkFilters(tag, msg, Log.WARN)
    return result
  }

  fun e(tag: String, msg: String): Int {
    val result = Log.e(tag, msg)
    checkFilters(tag, msg, Log.ERROR)
    return result
  }

  fun e(tag: String, msg: String, tr: Throwable): Int {
    val result = Log.e(tag, msg, tr)
    checkFilters(tag, msg, Log.ERROR)
    return result
  }

  fun wtf(tag: String, msg: String): Int {
    Log.wtf(tag, msg)
    checkFilters(tag, msg, Log.ASSERT)
    return 0
  }

  fun wtf(tag: String, msg: String, tr: Throwable): Int {
    Log.wtf(tag, msg, tr)
    checkFilters(tag, msg, Log.ASSERT)
    return 0
  }

  private fun checkFilters(tag: String, message: String, level: Int) {
    if (filters.isEmpty()) return
    val buf = buffer ?: return

    // First-match semantics — only record once
    for (filter in filters) {
      if (filter.matches(tag, message, level)) {
        buf.add(
          SdkLogEvent(
            timestamp = System.currentTimeMillis(),
            applicationId = applicationId,
            level = level,
            tag = tag,
            message = message,
            filterName = filter.name,
          )
        )
        return
      }
    }
  }
}

File: ios/auto-mobile-sdk/Sources/AutoMobileSDK/Logging/AutoMobileLog.swift
import Foundation
import os.log

/// Log filtering and recording.
/// Drop-in companion for os.Logger that records matching log entries to the event buffer.
public final class AutoMobileLog: @unchecked Sendable {
    public static let shared = AutoMobileLog()

    private let lock = NSLock()
    private var filters: [String: LogFilter] = [:]
    private var bundleId: String?
    private weak var buffer: SdkEventBuffer?

    private init() {}

    func initialize(bundleId: String?, buffer: SdkEventBuffer) {
        lock.lock()
        self.bundleId = bundleId
        self.buffer = buffer
        lock.unlock()
    }

    // MARK: - Filter Management

    /// Add a log filter. Logs matching the filter's patterns are recorded to the event buffer.
    public func addFilter(
        name: String,
        tagPattern: NSRegularExpression? = nil,
        messagePattern: NSRegularExpression? = nil,
        minLevel: LogLevel = .verbose
    ) {
        lock.lock()
        defer { lock.unlock() }
        filters[name] = LogFilter(
            name: name,
            tagPattern: tagPattern,
            messagePattern: messagePattern,
            minLevel: minLevel
        )
    }

    /// Convenience: add a filter with string regex patterns.
    /// Returns false and does not register the filter if a pattern is invalid.
    @discardableResult
    public func addFilter(
        name: String,
        tagPattern: String? = nil,
        messagePattern: String? = nil,
        minLevel: LogLevel = .verbose
    ) -> Bool {
        var tagRegex: NSRegularExpression?
        var messageRegex: NSRegularExpression?

        if let pattern = tagPattern {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
            tagRegex = regex
        }
        if let pattern = messagePattern {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
            messageRegex = regex
        }

        addFilter(name: name, tagPattern: tagRegex, messagePattern: messageRegex, minLevel: minLevel)
        return true
    }

    /// Remove a filter by name.
    public func removeFilter(name: String) {
        lock.lock()
        defer { lock.unlock() }
        filters.removeValue(forKey: name)
    }

    /// Remove all filters.
    public func clearFilters() {
        lock.lock()
        defer { lock.unlock() }
        filters.removeAll()
    }

    /// Number of active filters.
    public var filterCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return filters.count
    }

    // MARK: - Log Methods

    public func v(_ tag: String? = nil, _ message: String) {
        log(level: .verbose, tag: tag, message: message)
    }

    public func d(_ tag: String? = nil, _ message: String) {
        log(level: .debug, tag: tag, message: message)
    }

    public func i(_ tag: String? = nil, _ message: String) {
        log(level: .info, tag: tag, message: message)
    }

    public func w(_ tag: String? = nil, _ message: String) {
        log(level: .warning, tag: tag, message: message)
    }

    public func e(_ tag: String? = nil, _ message: String) {
        log(level: .error, tag: tag, message: message)
    }

    public func fault(_ tag: String? = nil, _ message: String) {
        log(level: .fault, tag: tag, message: message)
    }

    // MARK: - Internal

    private func log(level: LogLevel, tag: String?, message: String) {
        guard AutoMobileSDK.shared.isEnabled else { return }

        lock.lock()
        guard !filters.isEmpty else {
            lock.unlock()
            return
        }
        let currentFilters = Array(filters.values)
        let currentBuffer = buffer
        lock.unlock()

        for filter in currentFilters {
            if filter.matches(level: level, tag: tag, message: message) {
                let event = SdkLogEvent(
                    level: level,
                    tag: tag,
                    message: message,
                    filterName: filter.name
                )
                currentBuffer?.add(event)
            }
        }
    }

    // MARK: - Testing Support

    internal func reset() {
        lock.lock()
        filters.removeAll()
        bundleId = nil
        buffer = nil
        lock.unlock()
    }
}

// MARK: - LogFilter

struct LogFilter: Sendable {
    let name: String
    let tagPattern: NSRegularExpression?
    let messagePattern: NSRegularExpression?
    let minLevel: LogLevel

    func matches(level: LogLevel, tag: String?, message: String) -> Bool {
        guard level >= minLevel else { return false }

        if let tagPattern = tagPattern {
            guard let tag = tag else { return false }
            let range = NSRange(tag.startIndex..., in: tag)
            guard tagPattern.firstMatch(in: tag, range: range) != nil else { return false }
        }

        if let messagePattern = messagePattern {
            let range = NSRange(message.startIndex..., in: message)
            guard messagePattern.firstMatch(in: message, range: range) != nil else { return false }
        }

        return true
    }
}
